### ISSUE TITLE
LoRA finetune on Libero Spatial yields 0% success rate

### ISSUE THREAD

**Issue by Lastsource** [openvla/openvla#299]
Hello, I am trying to fine-tune OpenVLA on the Libero Spatial dataset following the instructions in the LoRA finetune README. I have converted the dataset as described, and I checked that the dataset_statistics.json file is similar to the “no_noop” version.

I used a batch size of 16 for finetuning. However, when I evaluate the model on the Libero Spatial tasks, the measured success rate is 0%, even though the rollout videos show that the robot moves but fails to complete the tasks.

Steps I followed:
	1.	Converted the Libero Spatial dataset according to the README.
	2.	Verified the dataset_statistics.json.
	3.	Finetuned using LoRA with batch size 16.
	4.	Evaluated using run_libero_eval.py with the checkpoint.

Is there anything I might be missing in the conversion, preprocessing, or LoRA finetuning procedure that could lead to zero success on Libero Spatial? Any guidance would be greatly appreciated!

**Comment by wanlh99** [comment]
me too. Has this problem been solved?

**Comment by gaogehan** [comment]
Me too, Have you addressed this problem? thank you!

**Comment by v-xchen-v** [comment]
Me too. For more details:
1. When evaluating the base model, the eef almost don't move. It make sense since it haven't see Libero-Spatial.
2. When evaluating the lora finetuned model, the eef move a little but not in correct directly and drop to the table after a while. And of course, 0 success rate. 
3. The different betweem my exp and the readme guide, is I decrease batchsize from 16 to 4. But since I used 4 GPU with --nproc-per-node 4, it shouldn't influence a lot. 
4. I trained 8k steps, and the training curve is weird as well:

<img width="1437" height="516" alt="Image" src="https://github.com/user-attachments/assets/bd135407-cfcf-4e07-aa37-32c087eef59b" />

**Comment by v-xchen-v** [comment]
> Me too. For more details:
> 
> 1. When evaluating the base model, the eef almost don't move. It make sense since it haven't see Libero-Spatial.
> 2. When evaluating the lora finetuned model, the eef move a little but not in correct directly and drop to the table after a while. And of course, 0 success rate.
> 3. The different betweem my exp and the readme guide, is I decrease batchsize from 16 to 4. But since I used 4 GPU with --nproc-per-node 4, it shouldn't influence a lot.
> 4. I trained 8k steps, and the training curve is weird as well:
> 
> <img alt="Image" width="1437" height="516" src="https://private-user-images.githubusercontent.com/8065765/533209456-bd135407-cfcf-4e07-aa37-32c087eef59b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc5NDgyNTEsIm5iZiI6MTc2Nzk0Nzk1MSwicGF0aCI6Ii84MDY1NzY1LzUzMzIwOTQ1Ni1iZDEzNTQwNy1jZmNmLTRlMDctYWEzNy0zMmMwODdlZWY1OWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDlUMDgzOTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmEwYjg0OWJkMjVkOTMwN2IxM2ZiNGQ4ZjY4MWVlMjI3NDc0Y2M3NzA3ZTMxNDY0YmVmOWRmNjY3NmE3NDI0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EXkquTwZPeWDK9yRF6TtyJvGSurhIcV3OMfKQSMSOgo">

I reduced the learning_rate, since I reduced batch size, then I got the much better training curve, and the eval video is much better as well.

**Comment by HaochenZ11** [comment]
Is there a fix for this? I did LoRA finetune on libero-object with the default script and action accuracy reaches 100% with loss close to 0%. However, when I eval on libero-object the success is 0 and the output actions just move the robot arm forward and away from all object.

**Comment by shant0602** [comment]
Same issue here — LoRA fine-tune on `libero_spatial_no_noops`, 10k steps, batch=16, LR=5e-4, 0% eval success with mode collapse (identical action tokens every step regardless of observation).

We verified the **eval environment is not the problem**: tested `openvla/openvla-7b-finetuned-libero-spatial` on the exact same setup (ARM64/GH200, osmesa, all patches) and got ~75% task success, consistent with the paper. So the issue is isolated to training.

Our hypothesis and full details: #333

The `v-xchen-v` finding about LR reduction is interesting — may be related to effective LR scaling, or may interact with the missing `clip_grad_norm_` in the optimizer step.

**Comment by shant0602** [comment]
**Root cause identified and fixed — see PR #334**

Following up on my earlier comment: the root cause is confirmed. `vla-scripts/finetune.py` is missing `clip_grad_norm_` before `optimizer.step()`. A single outlier batch produces an unclamped gradient large enough to permanently corrupt the LoRA weights.

**Fix (one line, PR #334):**
```python
torch.nn.utils.clip_grad_norm_(trainable_params, max_norm=1.0)
optimizer.step()
```

**Results with fix applied (LIBERO-Spatial, 10 tasks × 5 trials):**

| | Without fix | With fix |
|---|---|---|
| Train loss (10k steps) | 2.66 (spike at step ~6k) | converges cleanly |
| Action acc (training) | 0.39 | 0.82 (at 20k steps) |
| LIBERO-Spatial eval | 0% (mode collapse) | **72%** |

Full analysis and loss curves: #333
One-line PR: #334
Public 20k checkpoint: https://huggingface.co/shant0602/openvla-7b-libero-spatial-lora-20k
